### PR TITLE
Add basic support for white-listing tests to run on new PRs.

### PR DESCRIPTION
**Pull request by samoht**
Currently, the only working option is to say "test this" to allow
a PR to be tested.

Signed-off-by: Thomas Gazagnaire <thomas@gazagnaire.org>

### CODE REVIEW

**Comment by samoht** [comment]
/cc @dave-tucker The workflow for testing stuff is very basic atm but it should be able to do the basic things needed, e.g. build/test will be done if someone on the whitelist opened the PR or if someone on the whitelist said "test me". The code compiles but is untested ...

Before merging, a `whitelist` file should added at the root of this repo, containing GitHub logins (one per line). DataKit will monitor this repo and will recompute PRs "testability" everytime a new commit is done to this repo.

File: src/PR_state.ml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  let contains x body =
+    let re = Re.compile Re.(seq [rep space; str x; rep space]) in
+    Re.execp re body
+  in
+  if in_whitelist (PR.owner pr) then `InWhiteList
+  else
+    let test_this c =
+      (* The user doing the comment is in the whitelist and he/she
+         says "test this" *)
+      in_whitelist (Comment.user c) && contains "test this" (Comment.body c)
```
Comment (talex5) [inline src/PR_state.ml]: Looks like Jenkins uses `test this please` - maybe that's less prone to mistakes? (e.g. `Did you test this?`)

File: src/linuxkit_build.ml [hunk 2]
```diff
@@ -368,7 +381,7 @@ module Builder = struct
     | Error `Not_file -> DK.Tree.read_file tree Cache.Path.value >>*= load_json t
     | Error e -> Utils.failf "Unexpected DB error: %a" DK.pp_error e
 
-  let branch _t { Key.src; target} =
+  let branch _t { Key.src; target; _} =
```
Comment (talex5) [inline src/linuxkit_build.ml]: It doesn't make sense to ignore something that's in the key. If you want all PRs, branches and tags to rebuild whenever the whitelist changes, the (hash of) the whitelist needs to be in the branch name.

If you don't, then the whitelist should be in the context.

Most likely, checking whether someone is whitelisted has nothing to do with building LinuxKit and should be checked as a separate (non-cached) build step. Then, changing the whitelist will only rebuild PRs affected by the change.

File: src/linuxkit_build.ml [hunk 3]
```diff
@@ -385,14 +398,13 @@ module Result_cache = Cache.Make(Builder)
...
   let open! Term.Infix in
   Term.job_id >>= fun job_id ->
-  let key = { Builder.Key.
-              src;
-              target;
-            } in
-  Result_cache.find t job_id key
+  Term.target target >>= fun target_v ->
+  let key = { Builder.Key. src; target; whitelist; } in
+  let context = { Builder.job_id; target = target_v } in
```
Comment (talex5) [inline src/linuxkit_build.ml]: I don't think putting the target in the context makes sense either. Whether a PR is built does depend on the state of the PR, not just on the commit.

**Review by talex5 — COMMENTED** [review]
I think the whitelist check should be a separate (non-cached) build step. This would avoid duplicating the check in build and test, and should make it rebuild exactly when needed.

**Comment by samoht** [comment]
As suggested by @talex5, I've added the PR check as a separate pipeline step so it could be cached more easily. The current pipeline becomes:

- pull whitelist from `linuxkit/linuxkit-ci/master/whitelist`
- check that the target (PR, ref) can be built using the information stored in the whitelist
- build and test as it is now

**Comment by dave-tucker** [comment]
LGTM and a whitelist file now exists in this repo. Is this good to merge/test?

**Comment by samoht** [comment]
It is using unreleased features in datakit-github, so I need to add a few pins in the Dockerfile (doing that now). But if you managed to get it compiled it should be good to test.

**Comment by samoht** [comment]
I've just fixed the Dockerfile so you should be able to test it more easily.

**Comment by samoht** [comment]
What's the current plan regarding LinuxKit CI? Should I close this PR?

**Comment by justincormack** [comment]
Hmm, we don't have a very determinate plan over CI at present... so maybe.

**Pull request closed without merging**